api/src/main/java/org/pmiops/workbench/api/DataSetController.java

```diff
@@ -224,30 +229,66 @@ public DomainValuePair apply(DataSetValues dataSetValue) {
-              .map(value -> new DataSetPreviewValueList().value(value.getValue())).collect(Collectors.toList());
-
-      queryResponse.getValues().forEach(fieldValueList -> {
-        IntStream.range(0, fieldValueList.size()).forEach(columnNumber -> {
-          try {
-            valuePreviewList.get(columnNumber).addQueryValueItem(
-                fieldValueList.get(columnNumber).getValue().toString());
-          } catch (NullPointerException ex) {
-            log.severe(
-                String.format("Null pointer exception while retriving value for query: Column %s ", columnNumber));
-            valuePreviewList.get(columnNumber).addQueryValueItem("");
+      int retry = 0, rowsRequested = NO_OF_PREVIEW_ROWS ;
+      List<DataSetPreviewValueList> valuePreviewList = new ArrayList<>();
+
+      do {
+        try {
+          String query = queryJobConfiguration.getQuery().concat(" LIMIT "+ rowsRequested);
+
+          queryJobConfiguration = queryJobConfiguration.toBuilder().setQuery(query).build();
+
+          TableResult queryResponse = bigQueryService.executeQuery(bigQueryService
+              .filterBigQueryConfig(queryJobConfiguration), 60000l);
+          queryResponse.getSchema().getFields().forEach(fields -> {
+            valuePreviewList.add(new DataSetPreviewValueList().value(fields.getName()));
+          });
+
+          queryResponse.getValues().forEach(fieldValueList -> {
+            IntStream.range(0, fieldValueList.size()).forEach(columnNumber -> {
+              try {
+                valuePreviewList.get(columnNumber).addQueryValueItem(
+                    fieldValueList.get(columnNumber).getValue().toString());
+              } catch (NullPointerException ex) {
+                log.severe(
+                    String.format("Null pointer exception while retriving value for query: Column %s ", columnNumber));
+                valuePreviewList.get(columnNumber).addQueryValueItem("");
+              }
+            });
+          });
+          queryResponse.getSchema().getFields().forEach(fields -> {
+            DataSetPreviewValueList previewValue = valuePreviewList.stream()
+                .filter(preview -> preview.getValue().equalsIgnoreCase(fields.getName())).findFirst().get();
+            if (fields.getType() == LegacySQLTypeName.TIMESTAMP) {
+              List<String> queryValues = new ArrayList<String>();
+              DateFormat dateFormat = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss");
+              previewValue.getQueryValue().forEach(value -> {
+                try {
+                  Double fieldValue = Double.parseDouble(value);
+                  queryValues.add(dateFormat.format(new Date(fieldValue.longValue())));
+                } catch (NumberFormatException ex) {
+                  queryValues.add("");
+                }
+              });
+              previewValue.setQueryValue(queryValues);
+            }
+          });
+          retry = 2;
```

Comment:
You may be able to use `break`, rather than setting retry = 2. I think that would be cleaner if we decided to do more or less retries, not needing to change as often.